Use beacon state interfaces across

terencechain opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
-->

**What type of PR is this?**

> Feature


**What does this PR do? Why is it needed?**

Using interfaces defined in #8544 for beacon state arguments. This is to get us started with HF1

**Which issues(s) does this PR fix?**

N/A

**Other notes for review**

N/A

rauljordan commented:
Blocking until 3 approves

nisdas commented on beacon-chain/core/epoch/precompute/new.go:
```diff
@@ -17,9 +16,13 @@ import (
+func New(ctx context.Context, state iface.BeaconState) ([]*Validator, *Balance, error) {
 	ctx, span := trace.StartSpan(ctx, "precomputeEpoch.New")
 	defer span.End()
+	if state == nil || state.InnerStateUnsafe() == nil {
```
same as what we discussed earlier, the interface cannot be nil. So the check will not check for anything.

nisdas reviewed (commented):
Main thing that we would need to look out for is nil check on the state. (Whereas the check for a pointer would be positive the same isn't the case for an interface.

rkapka commented on beacon-chain/powchain/service_test.go:
```diff
@@ -429,10 +429,12 @@ func TestInitDepositCache_OK(t *testing.T) {
 	}
+	gs, _ := testutil.DeterministicGenesisState(t, 1)
 	beaconDB := dbutil.SetupDB(t)
 	s := &Service{
-		chainStartData: &protodb.ChainStartData{Chainstarted: false},
-		beaconDB:       beaconDB,
+		chainStartData:  &protodb.ChainStartData{Chainstarted: false},
+		beaconDB:        beaconDB,
+		preGenesisState: gs,
```
This change seems unrelated to the PR. Does the original test fail after the refactor?

terencechain replied:
Rationale: https://github.com/prysmaticlabs/prysm/pull/8552#issuecomment-791529031

rkapka commented on beacon-chain/state/getters.go:
```diff
@@ -568,6 +568,9 @@ func (b *BeaconState) NumValidators() int {
 	}
 	b.lock.RLock()
 	defer b.lock.RUnlock()
+	if b.state.Validators == nil {
+		return 0
+	}
```
This change should be in its own PR.

rkapka commented on beacon-chain/sync/pending_attestations_queue_test.go:
```diff
@@ -137,7 +137,7 @@ func TestProcessPendingAtts_HasBlockSaveUnAggregatedAtt(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, 1, len(atts), "Did not save unaggregated att")
-	assert.DeepEqual(t, att, atts[0], "Incorrect saved att")
+	assert.DeepSSZEqual(t, att, atts[0], "Incorrect saved att")
```
This change should be in another PR.

rkapka reviewed (commented):
I would like to see a few things extracted out to a separate PR. Because the diff here is large, I would avoid non-essential refactoring.

terencechain commented:
@nisdas @rkapka 

Feedbacks addressed. I removed all the nil checks and fixed the tests. The issue was we had tests that depends solely on `hasInnerState` to proceed, where b is defined as `nil` in tests.
```go
func (b *BeaconState) hasInnerState() bool {
	return b != nil && b.state != nil
}
```  

Now the tests had to be adapted to `(*stateTrie.BeaconState)(nil)` instead

rkapka commented:
Why is this PR labeled as blocked?

terencechain commented:
> Why is this PR labeled as blocked?

https://github.com/prysmaticlabs/prysm/pull/8552#issuecomment-790929064

terencechain commented:
Removing blocked. This PR has gotten 4 approvals so far. Unfortunately conflict is adding up and we could be reviewing the commit diffs since last approval
